### ISSUE TITLE
Improve support of property existance checking

### ISSUE THREAD

**Issue by mgtitimoli** [facebook/flow#7619]
Flow version: ALL

## Expected behavior
- `if (prop in obj) console.log(obj[prop])` should work
- `if (typeof obj[prop] !== "undefined") console.log(obj[prop])` should work
- `if (obj[prop] !== undefined) console.log(obj[prop])` should work
- `if (obj[prop]) console.log(obj[prop])` should work

## Actual behavior
There is no way to check if a property exist without getting an error

* Link to Try-Flow or Github repo: [Try Flow](https://flow.org/try/#0MYewdgzgLgBATgUwI4FcHQMpQIZRRdGAXhgG8AoGGASwBMAbBALhgAoAiOx9lzhhdgEoANJRgAzbNUa0WHSdIS0eMdgplDRVCCmDAESuex16Dy3if1Kh5AL4BucuSgBPAA4IYAJWRpMOPAhiGAASADVsej8AHlcPEHF4X3QoLFx8dAA+exgnUEhYAFt0CGwAc0ISCioAbURUFLTA9AA6dSUAXV4fBugYAAtsIPblLRg65P90gggWyzMu1R6-WEGg+etRBzzwPvqVpvwWZcaA-GDWFDBaBHFqMEMYbDAXQUdyakTWfdPpmjAYMUIKUKhBBGQxPkICBGC16CAyqwgSD0BNeqkzhAOm87E5Pmw4ggEoCSuVUT8poEOjAAIREEjsK43O4PZTg6owKEwhBwhFI0mgtEHTHYxy2PFfZFkiBC35U2n0mBM273JTsyG7bm8xFSwUUjHTUW4j6SgXkyYGqnqqhc2HwnVmmX6w5YnG2IA)

**Comment by goodmind** [comment]
@mgtitimoli you don't have `idle` in object type

**Comment by goodmind** [comment]
https://flow.org/try/#0MYewdgzgLgBATgUwI4FcHQMpQIZRRdGAXhgG8AoGGASwBMAbBALhgAoAiOx9lzhhdgEoANJRgAzbNUa0WHSdIS0eMdgplDRVCCmDAESuex16Dy3if1Kh5AL4BucuSgBPAA4IYAJWRpMOPAhiGAASADVsej8AHlcPEHF4X3QoLFx8dAA+exgnUEhYAFt0CGwAc3QWUgAfMS4EAH4WaDhqMDKtCSkZZqhW9s7LM17+svJq22CKKgBtRFQUtMD0ADp1JQBdXh8F6BgAC2wg9eVOueT-dIIIFaHN7YvYQ6C707tHcny9+b9UgPwWAByeqA4KsFBgWgIcRtQwwbBgFyCD7URKsH6Lf5BNowYoQUoVCCCMhiL5FYJ4gnoc67P5XDak8AQECMFb0EBlViFZF2JyothxBAJXElcrUjGXQIbGAAQiIJHYEKhMLA1mJ0xgZNxFNFhJpvyW+AZVC+LIQbI5XJ5tj5aMpYog+sx9Nl8pgSuhsNo6sZBW1JHteoldKlvuZrPZnO5jht5H5XN14sehogGx9JqZ5IDicdwZTxs1TLNFqj1qAA

They all work except for `in` check

**Comment by mgtitimoli** [comment]
@goodmind you specified a type to messages making idle optional, that way for sure there is no gonna be an error.

The problem is that currently Flow does not provide a way to check the existance of a key on an object without a type, I would expect this not to be an error and just be another case of refinement.

**Comment by FireyFly** [comment]
What type would it refine to if you were allowed to test a property that isn't present in the object?

A `foo?: number,` property would be `number | undefined`, but if you test it with e.g. `if (obj.foo != null)` it should refine to just `number` inside the conditional. Defining it as optional is what *enables* the refinement, but without having specified the property you aren't allowed to access it at all (which helps catch errors like typos causing you to inadvertently test a non-existent property on the object)

**Comment by mgtitimoli** [comment]
@FireyFly have you seen my example?

Let me rephrase the question, If you have a set of 3 keys all typed, and you have an object which only have entries with 2 keys of the set, is there a way I can tell Flow by using a condition with the set that would result in a subset with only the keys that object has?

**Comment by thecotne** [comment]
@mgtitimoli you can only check for optional fields in object

because there is no need to check for properties that are not allowed by type

untyped objects are considered sealed and are not allowed to have any extra properties

so you don't need to check for those extra properties

**Comment by mgtitimoli** [comment]
@thecotne let's say I have a fn that receives a key that it's a disjoint of 3 possible values, and then I have an object which has 2 entries with 2 keys of the disjoint, I would like to test if the given key is in the object to do something. This is what the issue is about, and I believe it's a pretty common pattern.

**Comment by thecotne** [comment]
two solutions 
 1. [key casted to string](https://flow.org/try/#0FDBmFcDsGMBcEsD2kAEAbRBzA0gUwJ4DykhARgFa5woAUiFANCgNYEBcAzrAE7ySYBKFAG9gKFPFC1W+Canrkho8eOjIOiNLgB0GTHQoBtGQF0BYlAF9g14HrxESFKrBrCAJgENYngExsUX0smAHIvHwBGEPN7AmIySjg3cL8AoNCU32igA)
2. [object with type](https://flow.org/try/#0FAFwngDgpgBA8gIwFYwLwwN7BjAJgQxHwEYB+ALhgGcQAnASwDsBzAGmz0PwCYKZGArgFsEUWsAC+wYADMBjAMYh6Ae0YwANiuYBpKGDiNESKEpgAKFckrHWMANb6AlJg70ZFx2BhMYVpC5YODgKalQqGlAAdFrMlsgA2l4Auk4cUlLAsXoGRsimIOYYBETclNwSdgDkJSRVadn6hsYFRbVlMBXV7fVAA)

**Comment by Mr-Wallet** [comment]
I'd be interested in an idiomatic way to refine types when we're not in control of what arguments we get. Disjoint unions on some sort of `type` property are all well and good, but we don't always have the luxury. I don't really mind if it requires some extra annotation trickery, but I should be able to tell flow, "Hey I _know_ this prop isn't on one part of the union, _that's literally why I'm checking this prop;_ I'm trying to do a runtime refinement."

Right now I don't see a way to do that without jumping through a bunch of data-transformation hoops and variable-juggling that make the code much harder to read and maintain.
